File: examples/learn/service/ros_srv/python/ros_srv_node_creator.py
# Copyright (c) 2019 The Felicia Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import sys

import import_order_resolver  # Should be before import felicia_py
import felicia_py as fel
from felicia.core.protobuf.master_data_pb2 import NodeInfo
import felicia_py.command_line_interface as cli

from ros_srv_client_node import RosSrvClientNode
from ros_srv_server_node import RosSrvServerNode
from examples.learn.service.simple_service_flag import SimpleServiceFlag


def main():
    fel.felicia_init()

    delegate = SimpleServiceFlag()
    parser = cli.FlagParser()
    parser.set_program_name('ros_srv_node_creator')
    if not parser.parse(len(sys.argv), sys.argv, delegate):
        sys.exit(1)

    master_proxy = fel.master_proxy
    s = master_proxy.start()
    if not s.ok():
        print("{} {}.".format(cli.RED_ERROR, s), file=sys.stderr)
        sys.exit(1)

    node_info = NodeInfo()
    node_info.name = delegate.name_flag.value

    if delegate.is_server_flag.value:
        master_proxy.request_register_node(
            RosSrvServerNode, node_info, delegate)
    else:
        master_proxy.request_register_node(
            RosSrvClientNode, node_info, delegate)

    fel.main_thread.run()


if __name__ == '__main__':
    main()

File: examples/learn/service/ros_srv/cc/ros_srv_node_creator.cc
// Copyright (c) 2019 The Felicia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "felicia/core/felicia_init.h"
#include "felicia/core/master/master_proxy.h"
#include "felicia/core/thread/main_thread.h"

#include "examples/learn/service/ros_srv/cc/ros_srv_client_node.h"
#include "examples/learn/service/ros_srv/cc/ros_srv_server_node.h"
#include "examples/learn/service/simple_service_flag.h"

namespace felicia {

int RealMain(int argc, char* argv[]) {
  FeliciaInit();

  SimpleServiceFlag delegate;
  FlagParser parser;
  parser.set_program_name("ros_srv_node_creator");
  if (!parser.Parse(argc, argv, &delegate)) {
    return 1;
  }

  MasterProxy& master_proxy = MasterProxy::GetInstance();
  Status s = master_proxy.Start();
  if (!s.ok()) {
    std::cerr << kRedError << s << std::endl;
    return 1;
  }

  NodeInfo node_info;
  node_info.set_name(delegate.name_flag()->value());

  if (delegate.is_server_flag()->value()) {
    master_proxy.RequestRegisterNode<RosSrvServerNode>(node_info, delegate);
  } else {
    master_proxy.RequestRegisterNode<RosSrvClientNode>(node_info, delegate);
  }

  MainThread& main_thread = MainThread::GetInstance();
  main_thread.Run();

  return 0;
}

}  // namespace felicia

int main(int argc, char* argv[]) { return felicia::RealMain(argc, argv); }